aws/logs_monitoring/forwarder.py

```diff
@@ -25,92 +28,158 @@
     DD_PORT,
     DD_TRACE_INTAKE_URL,
     DD_FORWARD_LOG,
+    DD_RETRY_EVENTS,
     SCRUBBING_RULE_CONFIGS,
     INCLUDE_AT_MATCH,
     EXCLUDE_AT_MATCH,
+    RETRY_INTERVAL_SECONDS,
 )
 
 logger = logging.getLogger()
 logger.setLevel(logging.getLevelName(os.environ.get("DD_LOG_LEVEL", "INFO").upper()))
-trace_connection = TraceConnection(
-    DD_TRACE_INTAKE_URL, DD_API_KEY, DD_SKIP_SSL_VALIDATION
-)
 
 
-def forward(logs, metrics, traces):
-    """
-    Forward logs, metrics, and traces to Datadog in a background thread.
-    """
-    if DD_FORWARD_LOG:
-        _forward_logs(logs)
-
-    _forward_metrics(metrics)
-
-    if len(traces) > 0:
-        _forward_traces(traces)
-
-
-def _forward_logs(logs):
-    """Forward logs to Datadog"""
-    if logger.isEnabledFor(logging.DEBUG):
-        logger.debug(f"Forwarding {len(logs)} logs")
-    logs_to_forward = filter_logs(
-        [json.dumps(log, ensure_ascii=False) for log in logs],
-        include_pattern=INCLUDE_AT_MATCH,
-        exclude_pattern=EXCLUDE_AT_MATCH,
-    )
-    scrubber = DatadogScrubber(SCRUBBING_RULE_CONFIGS)
-    if DD_USE_TCP:
-        batcher = DatadogBatcher(256 * 1000, 256 * 1000, 1)
-        cli = DatadogTCPClient(DD_URL, DD_PORT, DD_NO_SSL, DD_API_KEY, scrubber)
-    else:
-        batcher = DatadogBatcher(512 * 1000, 4 * 1000 * 1000, 400)
-        cli = DatadogHTTPClient(
-            DD_URL, DD_PORT, DD_NO_SSL, DD_SKIP_SSL_VALIDATION, DD_API_KEY, scrubber
+class Forwarder(object):
+    def __init__(self, function_prefix):
+        self.trace_connection = TraceConnection(
+            DD_TRACE_INTAKE_URL, DD_API_KEY, DD_SKIP_SSL_VALIDATION
         )
+        self.storage = Storage(function_prefix)
+        self.retry_interval_seconds = RETRY_INTERVAL_SECONDS + randint(1, 100)
 
-    with DatadogClient(cli) as client:
-        try:
+    def forward(self, logs, metrics, traces):
+        """
+        Forward logs, metrics, and traces to Datadog in a background thread.
+        """
+        if DD_FORWARD_LOG:
+            self._forward_logs(logs)
+        self._forward_metrics(metrics)
+        self._forward_traces(traces)
+
+    def retry(self):
+        """
+        Retry forwarding logs, metrics, and traces to Datadog.
+        """
+        if not DD_RETRY_EVENTS:
+            return
+
+        for prefix in RetryPrefix:
+            self._retry_prefix(prefix)
+
+    def _retry_prefix(self, prefix):
+        if not self._can_retry(prefix):
```

Comment:
thought: Maybe add a log here. We are in the case where the `DD_RETRY_EVENTS` is enabled but we can't get the lock, so we are not retrying content.